ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcPipelinedInsertionsSpec.scala

```diff
@@ -0,0 +1,93 @@
+  self: AsyncFlatSpec with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (on PostgreSQL)"
+
+  it should "allow idempotent transaction insertions" in {
+    val key = "some-key"
+    val create @ (offset, tx) = txCreateContractWithKey(alice, key, Some("1337"))
+    val maybeSubmitterInfo = submitterInfo(tx)
+    val preparedInsert = prepareInsert(maybeSubmitterInfo, tx, CurrentOffset(offset))
+    for {
+      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
+      // Assume the indexer restarts after events insertion
+      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
+      _ <- ledgerDao.storeTransactionState(preparedInsert)
+      // Assume the indexer restarts after state insertion
+      _ <- store(create) // The whole transaction insertion succeeds
+      result <- ledgerDao.transactionsReader
+        .lookupFlatTransactionById(tx.transactionId, tx.actAs.toSet)
+    } yield {
+      inside(result.value.transaction) { case Some(transaction) =>
+        transaction.commandId shouldBe tx.commandId.get
+        transaction.offset shouldBe ApiOffset.toApiString(offset)
+        transaction.effectiveAt.value.seconds shouldBe tx.ledgerEffectiveTime.getEpochSecond
+        transaction.effectiveAt.value.nanos shouldBe tx.ledgerEffectiveTime.getNano
+        transaction.transactionId shouldBe tx.transactionId
+        transaction.workflowId shouldBe tx.workflowId.getOrElse("")
+        inside(transaction.events.loneElement.event.created) { case Some(created) =>
+          val (nodeId, createNode: NodeCreate[ContractId]) =
+            tx.transaction.nodes.head
+          created.eventId shouldBe EventId(tx.transactionId, nodeId).toLedgerString
+          created.witnessParties should contain only (tx.actAs: _*)
+          created.agreementText.getOrElse("") shouldBe createNode.coinst.agreementText
+          created.contractKey.value.sum.text.value shouldBe key
+          created.createArguments shouldNot be(None)
+          created.signatories should contain theSameElementsAs createNode.signatories
+          created.observers should contain theSameElementsAs createNode.stakeholders.diff(
+            createNode.signatories
+          )
+          created.templateId shouldNot be(None)
+        }
+      }
+    }
+  }
+
+  it should "not retrieve a transaction with an offset past the ledger end (when requested by single party)" in {
+    assertNoResponseFor(singleCreate)
+  }
+
+  it should "not retrieve a transaction with an offset past the ledger end (when requested by multiple parties)" in {
+    assertNoResponseFor(multiPartySingleCreate)
+  }
+
+  private def assertNoResponseFor(
+      offsetTx: (Offset, LedgerEntry.Transaction)
+  ): Future[Assertion] = {
+    val (offset, tx) = offsetTx
+    val maybeSubmitterInfo = submitterInfo(tx)
+    val preparedInsert = prepareInsert(maybeSubmitterInfo, tx, CurrentOffset(offset))
+    for {
+      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
```

Comment:
You could even write one more step:
```suggestion
      _ <- ledgerDao.storeTransactionEvents(preparedInsert)
      _ <- ledgerDao.storeTransactionState(preparedInsert)
```

Reply:
Done

Reply:
Done